core/jobs/batch_jobs/opportunity_management_jobs.py

```diff
@@ -35,25 +37,190 @@
+                question_count=skill_opportunity.question_count
+            )
+            skill_opportunity_model.update_timestamps()
+            return {
+                'status': 'SUCCESS',
+                'job_result': job_run_result.JobRunResult(stdout='SUCCESS'),
+                'model': skill_opportunity_model
+            }
+        except Exception as e:
+            return {
+                'status': 'FAILURE',
+                'job_result': job_run_result.JobRunResult(
+                    stderr='FAILURE: %s' % e),
+                'model': None
+            }
+
+    def count_unique_question_ids(
+        self,
+        question_skill_link_models:
+            list[list[question_models.QuestionSkillLinkModel]]
+        ) -> int:
+        """Counts the number of unique question ids.
+
+        Args:
+            question_skill_link_models: list[[list[QuestionSkillLinkModel]].
+                2D array of QuestionSkillLinkModels.
+
+        Returns:
+            int. The number of unique question ids.
+        """
+
+        question_ids = [link.question_id
+                        for link_list in question_skill_link_models
+                        for link in link_list]
+        return len(set(question_ids))
+
+    def run(self) -> beam.PCollection[job_run_result.JobRunResult]:
+        all_question_skill_link_models = (
+            self.pipeline
+            | 'Get all non-deleted QuestionSkillLinkModels' >> (
+                ndb_io.GetModels(
+                    question_models.QuestionSkillLinkModel
+                    .get_all(include_deleted=False))
+                )
+            | 'Group QuestionSkillLinkModels by skill id' >>
+                beam.GroupBy(lambda n: n.skill_id)
+        )
+
+        all_skill_models = (
+            self.pipeline
+            | 'Get all non-deleted SkillModels' >> (
+                ndb_io.GetModels(
+                    skill_models.SkillModel.get_all(include_deleted=False)))
+            | 'Get skill from model' >> beam.Map(
+                skill_fetchers.get_skill_from_model)
+            | 'Group Skills by skill id' >> beam.GroupBy(lambda m: m.id)
+        )
+
+        skills_with_question_counts = (({
+            'skills': all_skill_models,
+            'question_skill_links': all_question_skill_link_models
+        })
+            | 'Merge by skill id' >> beam.CoGroupByKey()
+        )
+
+        opportunities_results = (
+            skills_with_question_counts
+            | beam.Map(
+                lambda n:
+                self._create_skill_opportunity_model(
+                    opportunity_domain.SkillOpportunity( # type: ignore[no-untyped-call]
+                        skill_id=n[1]['skills'][0][0].id,
+                        skill_description=n[1]['skills'][0][0].description,
+                        question_count=self.count_unique_question_ids(
+                            n[1]['question_skill_links']
+                        ))))
```

Comment:
This is quite complicated I suggest moving this to a function.

Reply:
Just to clarify, which part of this are you suggesting we move?

Reply:
I think that the whole lambda should be transformed into a function.

Reply:
Done